blog/_posts/2016-09-10-biojulia2016-mid.md

```diff
@@ -0,0 +1,222 @@
+tuned so that users can use it in real analysis with large files. It is also
+feasible to read a [CRAM](http://www.ebi.ac.uk/ena/software/cram-toolkit) file
+combining the BAM reader and `samtools view` command.
+
+An experimental feature is parallel processing using multiple threads.
+Multi-threading support is introduced in Julia 0.5 and we use it to parallelize
+decompression of BAM files. Here is a simple benchmark script to show how
+much reading speed can be improved with multiple threads:
+
+    using Bio.Align
+
+    # Count the number of mapped records.
+    function countmapped(reader)
+        ret = 0
+        record = BAMRecord()
+        while !eof(reader)
+            # in-place reading
+            read!(reader, record)
+            if ismapped(record)
+                ret += 1
+            end
+        end
+        return ret
+    end
+
+    println(open(countmapped, BAMReader, ARGS[1]))
+
+`JULIA_NUM_THREADS` environment variable controls the number of worker threads.
+The result below shows that the elapsed time is almost halved using two threads:
+
+    ~/.j/v/Bio $ time julia countmapped.jl SRR1238088.sort.bam
+    28040186
+           29.27 real        28.64 user         0.66 sys
+    ~/.j/v/Bio $ env JULIA_NUM_THREADS=2 time julia countmapped.jl SRR1238088.sort.bam
+    28040186
+           17.40 real        32.31 user         0.63 sys
+
+
+## Package to handle BGZF files
+
+BGZF (Blocked GZip Format) is a gzip-compliant file format commonly used in
+bioinformatics. BGZF can be read using standard gzip tools but files in the
+format are compressed block by block and special metadata are added to index the
+compressed files for random access. BAM files are compressed in this file format
+and sequence alignments in a specific genomic region can be retrieved
+efficiently.  BGZFStreams.jl is a new package to handle BGZF files like usual
+I/O streams and it is built on top of our
+[Libz.jl](https://github.com/BioJulia/Libz.jl) package. Parallel decompression
+mentioned above is implemented in this package layer.
+ 
+    julia> using BGZFStreams
+
+    julia> stream = BGZFStream("/Users/kenta/.julia/v0.5/BGZFStreams/test/bar.bgz")
+    BGZFStreams.BGZFStream{IOStream}(<mode=read>)
+
+    julia> readstring(stream)
+    "bar"
+
+<https://github.com/BioJulia/BGZFStreams.jl>
+
+
+## Sequence demultiplexing tool
+
+Sequence demultiplexing is a technique to distinguish the origin of a sequence
+using its artificially-attached "barcode" sequence. This is often used at a
+preprocessing phase after [multiplexed
+sequence](http://www.illumina.com/technology/next-generation-sequencing/multiplexing-sequencing-assay.html),
+a common technique to sequence multiple samples simultaneously.  A barcode
+sequence, however, may be corrupted due to sequencing error, and we need to
+find the best matching barcode from a barcode set.  The demultiplexer algorithm
+implemented in Bio.jl is based on a trie-like data structure to efficiently
```

Comment:
"based on a trie-like data structure, and efficiently finds the optimal barcode from the prefix of a DNA sequence read in constant time" or something like that?